### PR TITLE
sorting component add

### PR DESCRIPTION
**Pull request by LinYunMo**
Re: https://github.com/cocos/cocos-engine/pull/11094
Re: https://github.com/cocos/3d-tasks/issues/12273
Re: https://github.com/cocos/3d-tasks/issues/12287

### Changelog

*

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

**Comment by LinYunMo** [comment]
steal wait the editor support

File: cocos/sorting/sorting-layers.ts [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    id: number;
+    name: string;
+    value: number;
+}
+
+const SortingLayer = {
+    default: 0,
+};
+
+export class SortingLayers {
```
Comment (SantyWang) [inline cocos/sorting/sorting-layers.ts:49]: Need API docs

File: cocos/sorting/sorting-layers.ts [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    private static nameMap = new Map<number, string>();
+    private static indexMap = new Map<number, number>();
+
+    public static Enum = Enum(SortingLayer);
```
Comment (SantyWang) [inline cocos/sorting/sorting-layers.ts:57]: Need API docs

File: cocos/sorting/sorting-layers.ts [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+        Enum.update(SortingLayers.Enum);
+        Enum.sortList(SortingLayers.Enum, (a, b) => SortingLayers.getLayerIndex(a.number) - SortingLayers.getLayerIndex(b.number));
+
+        if (EDITOR) {
+            const scene = director.getScene();
+            if (!scene) {
+                return;
+            }
+            scene.walk((node) => {
+                const sort = node.getComponent('cc.Sorting');
```
Comment (SantyWang) [inline cocos/sorting/sorting-layers.ts:155]: getComponent by Class

Comment (LinYunMo) [reply cocos/sorting/sorting-layers.ts:155]: Will circular dependency

File: cocos/sorting/sorting-layers.ts (hunk 1 above)
Comment (LinYunMo) [reply cocos/sorting/sorting-layers.ts:49]: done

File: cocos/sorting/sorting-layers.ts (hunk 2 above)
Comment (LinYunMo) [reply cocos/sorting/sorting-layers.ts:57]: done

File: cocos/sorting/sorting.ts [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    protected _sortingLayer = SortingLayers.Enum.default; // Actually saved id
+    @serializable
+    protected _sortingOrder = 0;
+
+    private _modelRenderer: ModelRenderer | null = null;
+
+    protected __preload () {
+        this._modelRenderer = this.getComponent('cc.ModelRenderer') as ModelRenderer;
+        if (!this._modelRenderer) {
+            console.warn(`node '${this.node && this.node.name}' doesn't have any ModelRenderer component, it will not work`);
```
Comment (SantyWang) [inline cocos/sorting/sorting.ts]:
```suggestion
            console.warn(`node '${this.node && this.node.name}' doesn't have any ModelRenderer component, this component will not work. please add ModelRenderer component first`);
```

**Review by SantyWang — APPROVED** [review]

**Pull request merged**